teps/0027-task-execution-status-at-runtime.md

```diff
@@ -0,0 +1,133 @@
+---
+title: task-exec-status-at-runtime
+authors:
+  - "@pritidesai"
+creation-date: 2020-10-15
+last-updated: 2020-10-15
+status: proposed
+---
+
+# TEP-0027: Task Execution Status at Runtime
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+A `finally` task in a `pipeline` is executed just before exiting the `pipeline`. `Pipeline` executes `finally` tasks
+after all `dag` tasks finish successfully or one of the `dag` task results in failure i.e `finally` tasks are always
+executed in parallel irrespective of success or failure of a `dag` task. Such `finally` tasks are lacking access to the
+execution status of the `dag` tasks at runtime to better design and process success scenarios v/s failure scenarios.
+
+This proposal is enabling a `finally` task to access the execution status of `dag` tasks at runtime to better design
+`pipeline` based on the status.
+ 
+This proposal also enables any `dag` task to access the execution status of other `dag` tasks.
+
+## Motivation
+
+* A `pipeline` author wants to send notification to an external system based on the execution status of the `dag` tasks. 
+* A `pipeline` author wants to rollback if the deployment fails in that environment/cluster/region/provider.
+
+### Goals
+
+The main goal of this proposal is to allow access to the execution status of a task within the `pipeline` at runtime.
+And design this approach in the simplest possible way, without any extra processing. 
+
+An additional but important goal of this proposal is identify the different execution states of a `pipelineTask` and
+how each of those states derived based on `taskRun` and/or `pipelineRun` status.
+
+### Non-Goals
+
+* This proposal is not to expose the execution state of `finally` tasks since its not required.
+ 
+* This proposal is not to expose the entire status of a `taskRun`/`pipelineRun` as metadata at runtime.
+
+## Proposal
+
+Introduce a `param` value `$(context.pipelineRun.Tasks.<pipelineTask>)` which resolves to one of the execution states:
```

Comment:
i wonder if we can structure the variable replacement path such that we can extend it easily. e.g. instead of `context.pipelineRun.Tasks.deployment`, maybe something like:

`context.pipelineRun.Tasks.deployment.conditions.succeeded` or `context.pipelineRun.Tasks.deployment.status` (both would leave us open to add more fields, e.g. we could add support for cancelled and skipped by adding something like `context.pipelineRun.Tasks.deployment.skipped`) (in a comment further down i want to suggest we don't include "skipped" yet)

Reply:
replacing context.pipelineRun.Tasks.deployment with `context.pipelineRun.Tasks.deployment.status` is reasonable.

We have been generally replacing `prefix` in the param validation routines, `prefix` here being `context.pipelineRun.Tasks`. There is no support for trimming suffix like `status` in this case. It's implementation details but achievable.

Reply:
We could make it more verbose and have it as explicit `context.pipelineRun.Tasks.deployment.conditions.succeeded` set to `true` or `false` 🤔  instead of enum `(Succeeded, Failed, None)` with `context.pipelineRun.Tasks.deployment.status` but how do we represent `None`? And it feels too verbose.